src/utils/units.cpp

```diff
@@ -0,0 +1,98 @@
+//========================================================================================
+// (C) (or copyright) 2024. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include "units.hpp"
+
+namespace ArtemisUtils {
+
+constexpr Real Msolar = 1.988416e33;
+constexpr Real AU = 1.495978707e13;
+
+Units::Units(ParameterInput *pin, std::shared_ptr<StateDescriptor> pkg) {
+  std::string unit_system_str = pin->GetOrAddString("artemis/units", "type", "scalefree");
+  if (unit_system_str == "scalefree") {
+    unit_system_ = UnitSystem::scalefree;
+  } else if (unit_system_str == "cgs") {
+    unit_system_ = UnitSystem::cgs;
+  } else {
+    PARTHENON_FAIL("Unit system not recognized! Valid choices are [scalefree, cgs]");
+  }
+
+  if (unit_system_ != UnitSystem::scalefree) {
+    std::string unit_specifier = pin->GetOrAddString("artemis", "unit_system", "base");
+    if (unit_specifier == "base") {
+      length_ = pin->GetReal("artemis", "length");
+      time_ = pin->GetReal("artemis", "time");
+      mass_ = pin->GetReal("artemis", "mass");
+    } else if (unit_specifier == "orbit") {
+      const Real total_mass = pin->GetReal("artemis", "total_mass"); // Solar masses
+      const Real reference_radius = pin->GetReal("artemis", "reference_radius"); // AU
+      mass_ = total_mass * Msolar;
+      length_ = reference_radius * AU;
+      parthenon::constants::PhysicalConstants<parthenon::constants::CGS> pc;
+      time_ = std::sqrt(4. * M_PI * M_PI / (pc.gravitational_constant * mass_) *
+                        std::pow(length_, 3)); // Orbital period
+    } else {
+      PARTHENON_FAIL("Unit specifier not recognized!");
+    }
+  }
+
+  // Remaining conversion factors
+  energy_ = std::pow(length_, 2) * mass_ * std::pow(time_, -2);
+  number_density_ = std::pow(length_, -3);
+  mass_density_ = mass_ * number_density_;
```

Comment:
Others that may be useful are:
- [ ] opacity
- [ ] energy density and/or pressure
- [ ] velocity
- [ ] whatever units `cv` is

Reply:
Done :+1: